livekit-agents/livekit/agents/utils/av_sync.py

```diff
@@ -0,0 +1,158 @@
+import asyncio
+import logging
+import time
+from collections import deque
+from typing import Optional, Union
+
+import livekit.agents.utils as utils
+from livekit import rtc
+
+logger = logging.getLogger(__name__)
+
+
+class AVSynchronizer:
+    """Synchronize audio and video capture.
+
+    Usage:
+        av_sync = AVSynchronizer(
+            audio_source=audio_source,
+            video_source=video_source,
+            video_fps=video_fps,
+        )
+
+        async for video_frame, audio_frame in video_generator:
+            await av_sync.push(video_frame)
+            await av_sync.push(audio_frame)
+    """
+
+    def __init__(
+        self,
+        *,
+        audio_source: rtc.AudioSource,
+        video_source: rtc.VideoSource,
+        video_fps: float,
+        video_queue_size_ms: float = 1000,
+        _max_delay_tolerance_ms: float = 300,
+    ):
+        self._audio_source = audio_source
+        self._video_source = video_source
+        self._video_fps = video_fps
+        self._video_queue_size_ms = video_queue_size_ms
+        self._max_delay_tolerance_ms = _max_delay_tolerance_ms
+
+        self._stopped = False
+
+        self._video_queue_max_size = int(
+            self._video_fps * self._video_queue_size_ms / 1000
+        )
+        self._video_queue = asyncio.Queue[rtc.VideoFrame](
+            maxsize=self._video_queue_max_size
+        )
+        self._capture_video_task = asyncio.create_task(self._capture_video())
+
+    async def push(self, frame: Union[rtc.VideoFrame, rtc.AudioFrame]) -> None:
```

Comment:
Should this wait to have at least one AudioFrame and one VideoFrame to start the capture? So we synchronize the start of the stream as well

Reply:
I think no, the agent can stream the video frames only if there is no audio at the beginning. The user only need to make sure when there is a audio for playing, the first audio frame is pushed at the same time with the corresponding video frame.

Reply:
we need to make the queue_size smaller enough for that case, e.g. 100ms or even less.